numpy/random/mtrand/mtrand.pyx

```diff
@@ -4653,6 +4673,191 @@ cdef class RandomState:
+        method : string, optional
+            Specify the algorithm that is used to generate the sample.
+            Must be 'count' (the default) or 'iterative'.  See the Notes
+            for a description of the methods.
+
+        Returns
+        -------
+        sample : ndarray
+            Sample drawn from the multivariate hypergeometric distribution.
+
+        See Also
+        --------
+        hypergeometric : Draw samples from the (univariate) hypergeometric
+            distribution.
+
+        Notes
+        -----
+        The two methods do not return the same sequence of samples.
+
+        The "count" algorithm is roughly equivalent to the following numpy
+        code::
+
+            choices = np.repeat(np.arange(len(colors)), colors)
+            selection = np.random.choice(choices, nsample, replace=False)
+            sample = np.bincount(selection, minlength=len(colors))
+
+        The "count" algorithm uses a temporary array with length
+        ``sum(colors)``.
+
+        The "iterative" algorithm generates a single sample by using repeated
+        calls to the univariate hypergeometric sampler.  It is roughly
+        equivalent to::
+
+            sample = np.zeros(len(colors), dtype=np.int)
+            # `remaining` is the cumulative sum of `colors` from the last
+            # element to the first; e.g. if `colors` is [3, 1, 5], then
+            # `remaining` is [9, 6, 5].
+            remaining = np.cumsum(colors[::-1])[::-1]
+            for i in range(len(colors)-1):
+                if nsample < 1:
+                    break
+                sample[i] = np.random.hypergeometric(colors[i], remaining[i+1],
+                                                     nsample)
+                nsample -= sample[i]
+            sample[-1] = nsample
+
+        For some cases, the 'iterative' algorithm may be significantly faster
+        than the 'count' algorithm.
+
+        .. versionadded:: 1.14.0
+
+        Examples
+        --------
+        >>> colors = [16, 8, 4]
+        >>> np.random.seed(123)
+        >>> np.random.multivariate_hypergeometric(colors, 6)
+        array([4, 1, 1])
+        >>> np.random.multivariate_hypergeometric(colors, 6, size=3)
+        array([[3, 1, 2],
+               [3, 2, 1],
+               [2, 3, 1]])
+        >>> np.random.multivariate_hypergeometric(colors, 6, size=(2, 2))
+        array([[[4, 1, 1],
+                [4, 2, 0]],
+
+               [[3, 1, 2],
+                [4, 2, 0]]])
+
+        """
+        cdef npy_intp num_colors, sample_size
+        cdef long total
+        cdef long *colors_ptr
+        cdef int result
+
+        if method not in ['count', 'iterative']:
+            raise ValueError('method must be "count" or "iterative".')
+
+        if nsample < 0:
+            raise ValueError("nsample must be nonnegative.")
+
+        bad_colors = False
+        try:
+            colors = np.require(colors, dtype=np.int_, requirements='C')
```

Comment:
This errors on my machine.

Reply:
I don't think this works on Windows as expected:

```
colors = np.array([1,3,4,5,2**60],dtype=np.int64)

np.require(colors, dtype=np.int_, requirements='C')
Out[12]: array([1, 3, 4, 5, 0])
```

Is setting out-of-range values to 0 expected?

Reply:
NVM, colors was already defined from matplotlib.

Reply:
@bashtage No, that's not the desired behavior.  A value that is too big to be handled correctly should result in a ValueError.

Reply:
Note that that's just `colors.astype(np.int32)`, which shows the same behaviour. Casting out-of-range values is implemention/un-defined in C, and using the `% 2**32` behavior is not surprising.

Reply:
I agree that it should raise.  I don't think require does raise though -- in the code above it silently converts.  In fact, 

```
np.require(np.array([np.pi]), dtype=np.int_, requirements='C')
Out[2]: array([3])
```

it also truncates floats.  Maybe you need `PyArray_FROM_OTF` like most other functions, which will raise TypeError iirc.

Reply:
This would be far faster as `np.ascontiguousarray(colors, dtype=int)`